crates/cargo-contract/src/cmd/verify.rs

```diff
@@ -0,0 +1,191 @@
+             However, the original contract was built using `{expected_rustc_version}`. Please\n\
+             install the correct toolchain (`rustup install {expected_rustc_version}`) and\n\
+             re-run the `verify` command.",);
+        anyhow::ensure!(rustc_matches, mismatched_rustc.bright_yellow());
+
+        let expected_wasm_opt_version = build_info.wasm_opt_settings.version;
+        let keep_debug_symbols = build_info.wasm_opt_settings.keep_debug_symbols;
+        let handler = crate::wasm_opt::WasmOptHandler::new(
+            build_info.wasm_opt_settings.optimization_passes,
+            keep_debug_symbols,
+        )?;
+        let wasm_opt_version = handler.version();
+
+        let wasm_opt_matches = wasm_opt_version == expected_wasm_opt_version;
+        let mismatched_wasm_opt = format!(
+            "\nYou are trying to `verify` a contract using `wasm-opt` version `{wasm_opt_version}`.\n\
+             However, the original contract was built using `wasm-opt` version `{expected_wasm_opt_version}`.\n\
+             Please install the matching version and re-run the `verify` command.",
+        );
+        anyhow::ensure!(wasm_opt_matches, mismatched_wasm_opt.bright_yellow());
+
+        let args = ExecuteArgs {
+            manifest_path: manifest_path.clone(),
+            verbosity,
+            build_mode: build_info.build_mode,
+            network: Default::default(),
+            build_artifact: BuildArtifacts::CodeOnly,
+            unstable_flags: Default::default(),
+            optimization_passes: build_info.wasm_opt_settings.optimization_passes,
+            keep_debug_symbols,
+            skip_linting: true,
+            output_type: Default::default(),
+        };
+
+        let build_result = execute(args)?;
+
+        // 3. Grab the built Wasm contract and compare it with the Wasm from the metadata.
+        let reference_wasm = if let Some(wasm) = metadata.source.wasm {
+            wasm
+        } else {
+            anyhow::bail!(
+                "\nThe metadata for the reference contract does not contain a Wasm binary,\n\
+                therefore we are unable to verify the contract."
+                .to_string()
+                .bright_yellow()
+            )
+        };
+
+        let built_wasm_path = if let Some(wasm) = build_result.dest_wasm {
+            wasm
+        } else {
+            // Since we're building the contract ourselves this should always be populated,
+            // but we'll bail out here just in case.
+            anyhow::bail!(
+                "\nThe metadata for the workspace contract does not contain a Wasm binary,\n\
+                therefore we are unable to verify the contract."
+                .to_string()
+                .bright_yellow()
+            )
+        };
+
+        let fs_wasm = std::fs::read(built_wasm_path)?;
+        let built_wasm = SourceWasm::new(fs_wasm);
+
+        if reference_wasm != built_wasm {
```

Comment:
I believe we should also compare the contract ABI metadata itself? They should be identical too, and if not it might give clues as to what (if anything) is different.